### PR TITLE
Unit Dropship Assignment and Automatic Bay Loading

### PR DESCRIPTION
**Pull request by mkerensky**
Adds the ability to assign ground units/fighters/small craft to a dropship or warship bay on the TO&E page.
Assigned units can be automatically deployed to a scenario with the parent ship and then auto-loaded into that ship's transport bays when the scenario is started.

Also adds full ui and campaign save/load support for configuring Naval C3 networks

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 1]
```diff
@@ -1138,10 +1215,158 @@ public double getInsulatedCargoCapacity() {
...
+     */
+    public double getCorrectBayCapacity(int unitType, double unitWeight) {
+        final String METHOD_NAME = "getCorrectBayCapacity(unitType,unitWeight)"; //$NON-NLS-1$
+        switch (unitType) {
+            case UnitType.MEK:
+                return getCurrentMechCapacity();
+        case UnitType.AERO:
+        case UnitType.CONV_FIGHTER:
+            // Return a small craft slot if no ASF slots exist
+            if (getCurrentASFCapacity() > 0) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/campaign/unit/Unit.java]: Do we want just add these instead?

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 2]
```diff
@@ -1152,6 +1377,16 @@ public double getLightVehicleCapacity() {
...
         return bays;
     }
+    
+    // Get only bays to which a light tank has been assigned
+    public double getCurrentLightVehicleCapacity() {
+        return lVeeCapacity;
+    }
+    
+    // Used to assign a tank to a bay on a specific transport ship in the TO&E
+    public void setLightVehicleCapacity(double bays) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/campaign/unit/Unit.java]: I gotta ask, why double?

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java [hunk 3]
```diff
@@ -153,19 +225,56 @@ public void actionPerformed(ActionEvent action) {
...
+                StringJoiner cantLoad = new StringJoiner(", ");
+                String cantLoadReasons = StaticChecks.canTransportShipCarry(units, ship);
+                if (cantLoadReasons != null) {
+                    cantLoad.add(ship.getName() + " cannot load selected units for the following reasons: \n" +
+                            cantLoadReasons);
+                    //If the ship can't load the selected units, display a nag with the reasons why
+                    JOptionPane.showMessageDialog(null, cantLoad, "Warning", JOptionPane.WARNING_MESSAGE);
+                } else {
+                    //First, remove the units from any other Transport they might be on
+                    for (Unit u : units) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: I like that you can re-assign units without having to unassign them first.

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java [hunk 4]
```diff
@@ -242,20 +351,35 @@ public void actionPerformed(ActionEvent action) {
...
                             null,
                             "Are you sure you want to delete "
                                     + force.getFullName() + "?",
                                     "Delete Force?", JOptionPane.YES_NO_OPTION)) {
                         return;
                     }
+                    for (UUID id : force.getAllUnits()) {
+                        Unit unit = gui.getCampaign().getUnit(id);
+                        if (unit != null) {
+                            for (UUID oldShipId : unit.getTransportShipId().keySet()) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: This and the handler for REMOVE_UNIT seem to have identical logic for individual units. Let's unify that into a single worker function.

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java [hunk 5]
```diff
@@ -283,38 +407,71 @@ public void actionPerformed(ActionEvent action) {
...
             if(scenario instanceof AtBDynamicScenario) {
                 ForceTemplateAssignmentDialog ftad = new ForceTemplateAssignmentDialog(gui, null, units, (AtBDynamicScenario) scenario);
             } else {
+                HashSet<Unit> extraUnits = new HashSet<Unit>();
                 for (Unit unit : units) {
                     if (null != unit && null != scenario) {
+                        if (!unit.getTransportedUnits().isEmpty()) {
+                            // Prompt the player to also deploy any units transported by this one
+                            int optionChoice = JOptionPane.showConfirmDialog(null,
+                                    "You are deploying a Transport with units assigned to it. \n"
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: Let's tokenize this one.

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java [hunk 6]
```diff
@@ -970,18 +1145,53 @@ private void maybeShowPopup(MouseEvent e) {
...
-                menuItem.setActionCommand("REMOVE_FORCE|FORCE|empty|"
+                menuItem.setActionCommand(TOEMouseAdapter.COMMAND_REMOVE_FORCE
                         + forceIds);
                 menuItem.addActionListener(this);
                 menuItem.setEnabled(!StaticChecks.areAnyForcesDeployed(forces) && !StaticChecks.areAnyUnitsDeployed(unitsInForces));
                 popup.add(menuItem);
+                //Attempt to Assign all units in the selected force(s) to a transport ship. 
+                //This checks to see if the ship is in a basic state that can accept units. 
+                //Capacity gets checked once the action is submitted.
+                menu = new JMenu("Assign Force to Transport Ship");
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: Tokenize

File: MekHQ/src/mekhq/gui/dialog/ForceTemplateAssignmentDialog.java [hunk 7]
```diff
@@ -236,12 +237,28 @@ private void updateAssignButtonState() {
...
         Unit unit = unitList.getSelectedValue();
-        
         currentScenario.removeUnit(unit.getId());
         currentScenario.addUnit(unit.getId(), templateList.getSelectedValue().getForceName());
         unit.setScenarioId(currentScenario.getId());
         MekHQ.triggerEvent(new DeploymentChangedEvent(unit, currentScenario));
-        
+        if (!unit.getTransportedUnits().isEmpty()) {
+            // Prompt the player to also deploy any units transported by this one
+            int optionChoice = JOptionPane.showConfirmDialog(null,
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/dialog/ForceTemplateAssignmentDialog.java]: I feel like I've seen this exact same dialog elsewhere. Good candidate for a worker function if it's identical.

File: MekHQ/src/mekhq/gui/utilities/StaticChecks.java [hunk 8]
```diff
@@ -125,6 +246,72 @@ public static boolean areAllUnitsOnSameC3iNetwork(Vector<Unit> units) {
         }
         return true;
     }
+    
+    public static boolean doAllUnitsHaveNC3(Vector<Unit> units) {
```
Comment (NickAragua) [inline MekHQ/src/mekhq/gui/utilities/StaticChecks.java]: Comments on public methods.

**Comment by swang300** [comment]
One thing I didn't notice while scrolling through the changes is a change to make a loaded foot Infantry unit not count as a unit if the force size limitations are turned on. (ATB 2.31, Campaign_System A805)  While we're at it, I don't think the other 3 lines in that list is really implemented either.

File: MekHQ/src/mekhq/Utilities.java [hunk 9]
```diff
@@ -1595,4 +1604,107 @@ public static BufferedImage toBufferedImage(Image img) {
...
+            Entity cargo = client.getEntity(id);
+            if (cargo == null) {
+                continue;
+            }
+            // Find a bay with space in it and update that space so the next unit can process
+            cargo.setTargetBay(selectBestBayFor(cargo, transport));
+        }
+        // Reset transporter status again so that sendLoadEntity can process correctly
+        transport.resetTransporter();
+        for (int id : toLoad) {
```
[the remark this answers is not in the document]
Comment (mkerensky) [reply MekHQ/src/mekhq/Utilities.java]: This code is making sure that bays don't stop loading due to 'doors' being consumed, which isn't affected by clock timing.

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 10]
```diff
@@ -1027,13 +1030,42 @@ private void addUnit(Unit u) {
                 "Adding unit: (" + u.getId() + "):" + u); //$NON-NLS-1$
         units.put(u.getId(), u);
         checkDuplicateNamesDuringAdd(u.getEntity());
+        
+        //If this is a ship, add it to the list of potential transports
+        //Jumpships and space stations are intentionally ignored at present, because this functionality is being
+        //used to auto-load ground units into bays, and doing this for large craft that can't transit is pointless.
+        if (u.getEntity() != null && (u.getEntity() instanceof Dropship || u.getEntity() instanceof Warship)) {
```
[the remark this answers is not in the document]
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/Campaign.java]: I intentionally left them out too for the same reason. I'm going to allow you to load bay-carried infantry or assigned marines into small craft when you launch one from an MM game eventually, and also allow them to add to the escape pod capacity in some situations.

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 11]
```diff
@@ -268,9 +306,41 @@ public void run() {
...
+                    if (!scenario.getPlayerTransportLinkages().isEmpty()) {
+                        boolean loadFighters = false;
+                        boolean loadGround = false;
+                        for (UUID id : scenario.getPlayerTransportLinkages().keySet()) {
+                            Unit transport = campaign.getUnit(id);
+                            Set<Integer> toLoad = new HashSet<>();
+                            // Let the player choose to load fighters and/or ground units on each transport
+                            if (transport.isCarryingAero()) {
+                                loadFighters = (JOptionPane.YES_OPTION == JOptionPane.showConfirmDialog(null,
+                                                    "Would you like the fighters assigned to " + transport.getName()
```
[the remark this answers is not in the document]
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java]: Fixed.

File: MekHQ/src/mekhq/Utilities.java [hunk 12]
```diff
@@ -1595,4 +1604,107 @@ public static BufferedImage toBufferedImage(Image img) {
...
+            // And now load the units
+            if (cargo.isFighter() && loadFighters && transport.canLoad(cargo, false) && cargo.getTargetBay() != -1) {
+                client.sendLoadEntity(id, trnId, cargo.getTargetBay());
+            } else if(loadGround && transport.canLoad(cargo, false) && cargo.getTargetBay() != -1) {
+                client.sendLoadEntity(id, trnId, cargo.getTargetBay());
+            }
+        }
+    }
+    
+    public static int selectBestBayFor(Entity cargo, Entity transport) {
```
[the remark this answers is not in the document]
Comment (mkerensky) [reply MekHQ/src/mekhq/Utilities.java]: Fixed

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 13]
```diff
@@ -6738,6 +6833,34 @@ public void addUnitsToNetwork(Vector<Unit> addedUnits, String netid) {
         }
         return networks;
     }
+    
+    public Vector<String[]> getAvailableNC3Networks() {
```
[the remark this answers is not in the document]
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/Campaign.java]: Fixed

File: MekHQ/src/mekhq/campaign/unit/Unit.java (hunk 2 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/unit/Unit.java]: I was trying to preserve the datatype of capacity as defined in Bay, which is a double when used for Infantry and Cargo bays.

File: MekHQ/src/mekhq/campaign/unit/Unit.java (hunk 1 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/unit/Unit.java]: No I don't think we do.  I'm trying to allow for the possibility that we want to load ASFs onto a ship that has only small craft bays.  There are several of those, especially in the Pocket Warship class

File: MekHQ/src/mekhq/Utilities.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkerensky) [reply MekHQ/src/mekhq/Utilities.java]: Fixed.

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java (hunk 4 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: Fixed.

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java (hunk 5 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: Fixed

File: MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java (hunk 6 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java]: Fixed

File: MekHQ/src/mekhq/gui/dialog/ForceTemplateAssignmentDialog.java (hunk 7 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/gui/dialog/ForceTemplateAssignmentDialog.java]: Fixed

File: MekHQ/src/mekhq/gui/utilities/StaticChecks.java (hunk 8 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/gui/utilities/StaticChecks.java]: Fixed

**Pull request merged**